### PR TITLE
Early pull zip, concat and merge preferred stages

### PR DESCRIPTION
**Pull request by 2m**

### CODE REVIEW

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowJoinSpec.scala [hunk 1]
```diff
@@ -18,8 +22,11 @@ class FlowJoinSpec extends AkkaSpec(ConfigFactory.parseString("akka.loglevel=INF
 
   implicit val materializer = ActorMaterializer(settings)
 
+  implicit val defaultPatience =
+    PatienceConfig(timeout = Span(2, Seconds), interval = Span(200, Millis))
```
Comment (rkuhn) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowJoinSpec.scala]: You can also say `PatienceConfig(2.seconds)` if you don’t care about the interval (which I usually don’t).

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template [hunk 2]
```diff
@@ -56,12 +65,8 @@ class ZipWith1[[#A1#], O] (zipper: ([#A1#]) ⇒ O) extends GraphStage[FanInShape
...
-        if (willShutDown) completeStage()
-        else {
-            [#pull(in0)#
-            ]
-        }
+        pending += shape.inlets.size
+        if (pending == ##0) pushAll()
       }
     })
   }
```
Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: the toString could use some ZipWith1 love …

File: akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala [hunk 3]
```diff
@@ -158,15 +159,13 @@ final class MergePreferred[T] private (val secondaryPorts: Int, val eagerClose:
...
-      private var first = true
-      override def onPull(): Unit = {
-        if (first) {
-          first = false
-          tryPull(preferred)
-          shape.inSeq.foreach(tryPull)
-        }
-      }
+      override def onPull(): Unit = ()
     })
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala]: This is now `eagerTerminateOutput` (i.e. could save one allocation).

**Comment by rkuhn** [comment]
LGTM

**Comment by ktoso** [comment]
Failure was a compilation error. I can take this through the finish line, it is still relevant AFAICS right?

**Comment by rkuhn** [comment]
Yes, would be good; there were a few other comments as well.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala [hunk 4]
```diff
@@ -62,14 +62,13 @@ object GraphStages {
 
   def identity[T] = Identity.asInstanceOf[SimpleLinearGraphStage[T]]
 
-  private class Detacher[T] extends GraphStage[FlowShape[T, T]] {
+  private[stream] class Detacher[T] extends GraphStage[FlowShape[T, T]] {
```
Comment (viktorklang) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: `final`?

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/GraphStages.scala]: I'm taking it over, will add :)

File: akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: I don't quite get this comment hm hm

Comment (rkuhn) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: The toString method should include the number IIRC

Comment (ktoso) [inline akka-stream/src/main/boilerplate/akka/stream/scaladsl/ZipWithApply.scala.template]: Thanks!
Applying stuff now and will make it a shorter day tomorrow :-)

**Comment by ktoso** [comment]
Superseeded by https://github.com/akka/akka/issues/19270

**Comment by ktoso** [comment]
Reopening, @2m works on it.
@viktorklang had a comment or two on: https://github.com/akka/akka/pull/19270

**Comment by rkuhn** [comment]
@2m are you picking this up again?

**Comment by 2m** [comment]
Yes.

**Comment by rkuhn** [comment]
closed by github due to deletion of the release-2.3-dev branch—please open a new PR against master

**Pull request closed without merging**